File: 2022/10/day10.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INST_DELIM " "

int main() {
  FILE *problem_input = fopen("day10.prod", "r");

  int acc = 1;
  int cycle_num = 0;

  // partone
  int sig_strength = 0;
  // parttwo
  char crt_display[1001];

  char *token;
  int opcycles;
  int is_addx;
  for (char line[101]; fgets(line, sizeof(line), problem_input) != NULL;) {
    is_addx = 0;

    // first token is an instruction
    token = strtok(line, INST_DELIM);

    if (strncmp(token, "noop", 4) == 0) {
      opcycles = 1;
    } else if (strncmp(token, "addx", 4) == 0) {
      opcycles = 2;
      is_addx = 1;
    } else {
      printf("Something is wrong...");
      fclose(problem_input);
      exit(1);
    }

    for (; opcycles > 0; --opcycles) {
      ++cycle_num;

      /* partone */
      if (20 <= cycle_num && cycle_num <= 220 && (cycle_num + 20) % 40 == 0) {
        sig_strength += acc * cycle_num;
      }

      // DEBUG
      // printf("[%d]: %d ~ %d, %d\n", cycle_num, acc, sig_strength,
      //        acc * cycle_num);

      /* parttwo */
      if (0 <= cycle_num && cycle_num <= 240) {
        // position of the crt cursor: (cycle_num - 1) % 40
        // position of the sprite: acc

        if (abs((cycle_num - 1) % 40 - acc) < 2) {
          strcat(crt_display, "#");
        } else {
          strcat(crt_display, ".");
        }
        if ((cycle_num - 1) % 40 == 39)
          strcat(crt_display, "\n");
      }
    }

    if (is_addx) {
      // second token is an operand
      token = strtok(NULL, INST_DELIM);
      acc += atoi(token);
    }
  }

  fclose(problem_input);

  // partone: 16880
  printf("Signal strength: %d\n", sig_strength);

  // parttwo: RKAZAJBR
  printf("CRT display:\n%s", crt_display);

  /*

  ###..#..#..##..####..##....##.###..###..
  #..#.#.#..#..#....#.#..#....#.#..#.#..#.
  #..#.##...#..#...#..#..#....#.###..#..#.
  ###..#.#..####..#...####....#.#..#.###..
  #.#..#.#..#..#.#....#..#.#..#.#..#.#.#..
  #..#.#..#.#..#.####.#..#..##..###..#..#.

  */

  return 0;
}

File: 2022/10/day10.ts
// @ts-ignore TS, stop it already...
import { readFileSync } from "node:fs";

let instructions = (
  readFileSync("./day10.prod", { encoding: "utf-8" }) as string
)
  .split("\n")
  .map((line) => line.split(" ")) as string[][];

// pop final newline
instructions.pop();

let acc = 1;
let cycleNum = 0;

// partone
let signalStrength = 0;
//parttwo
let crtDisplay = "";

for (let inst of instructions) {
  let opcode = inst[0];
  let opcycles = 0;

  if (opcode == "noop") {
    opcycles = 1;
  } else if (opcode == "addx") {
    opcycles = 2;
  } else {
    console.log("Something is wrong...");
    break;
  }

  for (; opcycles > 0; --opcycles) {
    // it took me hours to figure out i had to put
    // this little bad boy specifically here
    ++cycleNum;

    //#region partone
    if (20 <= cycleNum && cycleNum <= 220 && (cycleNum + 20) % 40 === 0) {
      signalStrength += acc * cycleNum;
    }
    // console.log(cycleNum, acc, signalStrength, acc * cycleNum); // DEBUG
    //#endregion

    //#region parttwo
    if (0 <= cycleNum && cycleNum <= 240) {
      let crtCursorPos = (cycleNum - 1) % 40;
      let spritePos = acc;
      let areAroundSameSpot = Math.abs(crtCursorPos - spritePos) < 2;

      crtDisplay += areAroundSameSpot ? "#" : ".";
      if (crtCursorPos == 39) crtDisplay += "\n";
    }
    //#endregion
  }

  if (opcode === "addx") {
    let operand = inst[1];
    acc += parseInt(operand, 10);
  }
}

// partone: 16880
console.log("Signal strength: " + signalStrength);

// parttwo: RKAZAJBR
console.log(crtDisplay.trimEnd());

/*

###..#..#..##..####..##....##.###..###..
#..#.#.#..#..#....#.#..#....#.#..#.#..#.
#..#.##...#..#...#..#..#....#.###..#..#.
###..#.#..####..#...####....#.#..#.###..
#.#..#.#..#..#.#....#..#.#..#.#..#.#.#..
#..#.#..#.#..#.####.#..#..##..###..#..#.

*/
